tree-wide: add missing parent icon themes

romildo opened the pull request:
<!-- Nixpkgs has a lot of new incoming Pull Requests, but not enough people to review this constant stream. Even if you aren't a committer, we would appreciate reviews of other PRs, especially simple ones like package updates. Just testing the relevant package/service and leaving a comment saying what you tested, how you tested it and whether it worked would be great. List of open PRs: <https://github.com/NixOS/nixpkgs/pulls>, for more about reviewing contributions: <https://hydra.nixos.org/job/nixpkgs/trunk/manual/latest/download/1/nixpkgs/manual.html#chap-reviewing-contributions>. Reviewing isn't mandatory, but it would help out a lot and reduce the average time-to-merge for all of us. Thanks a lot if you do! -->
###### Motivation for this change

Icon themes may inherit from other icon themes. The inheritance is specified using the `Inherits` key in the `index.theme` file.

According to the [icon theme specification](https://specifications.freedesktop.org/icon-theme-spec/icon-theme-spec-latest.html) that means that icons not provided by the theme are looked for in its parents. The idea is that missing icons are looked for in the parent icon themes, in the order specified.

Therefore the parent themes should be installed as dependencies for a more complete experience regarding the icon sets used.

###### Things done

<!-- Please check what applies. Note that these are not hard requirements but merely serve as information for reviewers. -->

- [x] Tested using sandboxing ([nix.useSandbox](http://nixos.org/nixos/manual/options.html#opt-nix.useSandbox) on NixOS, or option `sandbox` in [`nix.conf`](http://nixos.org/nix/manual/#sec-conf-file) on non-NixOS linux)
- Built on platform(s)
   - [x] NixOS
   - [ ] macOS
   - [ ] other Linux distributions
- [ ] Tested via one or more NixOS test(s) if existing and applicable for the change (look inside [nixos/tests](https://github.com/NixOS/nixpkgs/blob/master/nixos/tests))
- [ ] Tested compilation of all pkgs that depend on this change using `nix-shell -p nixpkgs-review --run "nixpkgs-review wip"`
- [ ] Tested execution of all binary files (usually in `./result/bin/`)
- [ ] Determined the impact on package closure size (by running `nix path-info -S` before and after)
- [ ] Ensured that relevant documentation is up to date
- [x] Fits [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/.github/CONTRIBUTING.md).

Mic92 commented on pkgs/data/icons/papirus-maia-icon-theme/default.nix:
```diff
@@ -16,6 +16,12 @@ stdenv.mkDerivation rec {
+  buildInputs = [
+    breeze-icons
+    gnome-icon-theme
+    papirus-icon-theme
```
Do they get referenced in the build?

romildo replied:
No. They get referenced in the installed file `share/icons/Papirus-Adapta-Maia/index.theme` (and others):

```
Inherits=Papirus,breeze,gnome,hicolor
 ```

And it expects just that those themes are installed.

Mic92 replied:
Ok. The problem is that just adding them to `buildInputs` won't make them available at runtime or installed into the user's profile.
cc @jtojnar

jtojnar replied:
Yeah, I think the inherited themes need to go to `propagatedBuildInputs`. Recently encountered this issue in https://github.com/NixOS/nixpkgs/issues/84983.

jtojnar commented on pkgs/desktops/cinnamon/mint-y-icons/default.nix:
```diff
@@ -18,6 +19,7 @@ stdenv.mkDerivation rec {
 
   propagatedUserEnvPkgs = [
```
I think `propagatedBuildInputs` should be enough but I am not completely sure.

romildo replied:
This derivation was already using `propagatedUserEnvPkgs` and I have just added missing dependencies. I will change it to `propagatedBuildInputs`, handling the dependencies uniformly.

romildo replied:
Is `propagatedUserEnvPkgs` documented somewhere? I am not finding in on the [nixpkgs manual](https://nixos.org/nixpkgs/manual/).

jtojnar replied:
It is not documented because it is frowned upon.

But reading https://github.com/NixOS/nixpkgs/issues/34657#issuecomment-363465755, `propagatedBuildInputs` might not work.

romildo commented:
I have tested with `propagatedBuildInputs` and with `propagatedUserEnvPkgs` in a VM configured with the `startx` display manager and the `fluxbox` window manager. I have listed the files in `/run/current-system/sw/share/icons`. I have also looked at the themes with the application `lxappearance`. The only theme explicitly installed was `papirus-maia-icon-theme` (using `environment.systemPackages`).

Only with `propagatedUserEnvPkgs` the parent themes were made available automatically.

The following screenshots illustrates the results.

With `propagatedBuildInputs`:

![Screenshot_2020-04-14_20-47-56](https://user-images.githubusercontent.com/1217934/79284985-90547300-7e92-11ea-8941-b8200cbc7130.png)


With `propagatedUserEnvPkgs`:

![Screenshot_2020-04-14_20-51-47](https://user-images.githubusercontent.com/1217934/79284979-8a5e9200-7e92-11ea-84e6-8b66e2c92956.png)

I think this is conclusive.

jtojnar commented:
I think it might be preferable (even though ugly) to symlink the inherited themes into the output. Relying on `propagatedUserEnvPkgs` is just too big of a violation of encapsulation and purity for my taste.

Mic92 commented:
If symlink works, than this should be the preferred solution.
In future a setup-hook that scans `index.theme` performs the symlink on its own might be the way to go.

michaelpj commented:
> I think it might be preferable (even though ugly) to symlink the inherited themes into the output.

Won't this result in collisions if the parent theme is installed separately? (That might be fine, they're the same content after all).

romildo commented:
Testing with symbolic links in `papirus-maia-icon-theme` with both `papirus-maia-icon-theme` and `breeze-icons` in `environment.systemPackages`:

![Screenshot_2020-04-15_09-02-11](https://user-images.githubusercontent.com/1217934/79335552-95e4a400-7ef8-11ea-93f1-672dc6daa721.png)


<details><summary>Collisions are reported at build time:</summary>

```
$ nixos-rebuild build-vm --show-trace --option sandbox true -I nixpkgs=/alt/nixpkgs -I nixos-config=/alt/nixos/configuration.fluxbox.nix 2>&1 | tee /var/tmp/nixos-rebuild-vm.log
building Nix...
building the system configuration...
these derivations will be built:
  /nix/store/1cspvw79b958fsw7073d2cxw55ylj6bc-system-path.drv
[...]
building '/nix/store/1cspvw79b958fsw7073d2cxw55ylj6bc-system-path.drv'...
collision between `/nix/store/bq8hnk8n1xyy18nmwd7mf78kirs2s0d9-breeze-icons-5.68.0/share/icons/breeze-dark/actions/16@2x/stateshape.svg' and `/nix/store/89bsafcnz0a6r1wfaiifvmsnbh6ccnxz-papirus-maia-icon-theme-2019-07-26/share/icons/breeze-dark/actions/16@2x/stateshape.svg'
collision between `/nix/store/bq8hnk8n1xyy18nmwd7mf78kirs2s0d9-breeze-icons-5.68.0/share/icons/breeze-dark/actions/16@2x/dialog-object-properties.svg' and `/nix/store/89bsafcnz0a6r1wfaiifvmsnbh6ccnxz-papirus-maia-icon-theme-2019-07-26/share/icons/breeze-dark/actions/16@2x/dialog-object-properties.svg'
collision between `/nix/store/bq8hnk8n1xyy18nmwd7mf78kirs2s0d9-breeze-icons-5.68.0/share/icons/breeze-dark/actions/16@2x/document-save-all.svg' and `/nix/store/89bsafcnz0a6r1wfaiifvmsnbh6ccnxz-papirus-maia-icon-theme-2019-07-26/share/icons/breeze-dark/actions/16@2x/document-save-all.svg'
collision between `/nix/store/bq8hnk8n1xyy18nmwd7mf78kirs2s0d9-breeze-icons-5.68.0/share/icons/breeze-dark/actions/16@2x/kruler-south.svg' and `/nix/store/89bsafcnz0a6r1wfaiifvmsnbh6ccnxz-papirus-maia-icon-theme-2019-07-26/share/icons/breeze-dark/actions/16@2x/kruler-south.svg'
collision between `/nix/store/bq8hnk8n1xyy18nmwd7mf78kirs2s0d9-breeze-icons-5.68.0/share/icons/breeze-dark/actions/16@2x/highlight-pointer-spot.svg' and `/nix/store/89bsafcnz0a6r1wfaiifvmsnbh6ccnxz-papirus-maia-icon-theme-2019-07-26/share/icons/breeze-dark/actions/16@2x/highlight-pointer-spot.svg'
[...]
collision between `/nix/store/bq8hnk8n1xyy18nmwd7mf78kirs2s0d9-breeze-icons-5.68.0/share/icons/breeze/places/64/folder-downloads.svg' and `/nix/store/89bsafcnz0a6r1wfaiifvmsnbh6ccnxz-papirus-maia-icon-theme-2019-07-26/share/icons/breeze/places/64/folder-downloads.svg'
created 39920 symlinks in user environment
gtk-update-icon-cache: Cache file created successfully.
gtk-update-icon-cache: Cache file created successfully.
gtk-update-icon-cache: Cache file created successfully.
building '/nix/store/ylzwcnfziq0sh9h2cdcxdxq89s5i4app-dbus-1.drv'...
building '/nix/store/0ggf9zbx6ackvk98fqkyr4ccyy1lx747-unit-polkit.service.drv'...
building '/nix/store/q6n9wsd55dvpqnr3d3xh7fj1jprf8pzm-unit-systemd-fsck-.service.drv'...
building '/nix/store/93i3jmhp60v3xxavm22fn8d571ljy4dz-unit-dbus.service.drv'...
building '/nix/store/scjf25hizm5c8lyl9fdbsjnaaxclgxyx-unit-dbus.service.drv'...
building '/nix/store/f73af23ffjps3iw9l7m48k39fqm8dqir-system-units.drv'...
building '/nix/store/kwnp8n3jczl3wmim25mixbl178478xa8-user-units.drv'...
building '/nix/store/44rvwlz38pjra1bryds6qr0dy0svmlk0-etc.drv'...
building '/nix/store/1crsvb182anj0b9was6bqgdjxqs1yv9v-nixos-system-nixos-20.09pre-git.drv'...
building '/nix/store/rg3qf72c3xcvzygc16050jf68q7rrfsw-closure-info.drv'...
building '/nix/store/zxzpd7zcpmskib7yjcg3xi1h0havalkw-run-nixos-vm.drv'...
building '/nix/store/afqj6dvskdb2hf0962yiz76qv6mkm0mq-nixos-vm.drv'...

Done.  The virtual machine can be started by running /nix/store/088zz2r26xw8vj61kmrrcfcp3h173vzn-nixos-vm/bin/run-nixos-vm
```

</details>

romildo commented:
Should I go on and implement this solution for all icon themes?

romildo committed:
xfce.xfce4-icon-theme: add missing parent icon themes

Also:
- replace gtk2 by gtk3
- intltool and gtk are native build dependences

romildo committed:
numix-icon-theme-circle: add missing parent icon themes

Also:
- add runHook calls
- move postFixup actions to installPhase

romildo committed:
numix-icon-theme-square: add missing parent icon themes

Also:
- add runHook calls
- move postFixup actions to installPhase

romildo committed:
papirus-maia-icon-theme: add missing parent icon themes

Also:
- move postFixup actions to postInstallPhase

romildo committed:
yaru-theme: add missing parent icon themes

romildo committed:
cinnamon.mint-y-icons: add missing parent icon themes

Also:
- add runHook calls
- move postFixup actions to installPhase
- gtk-update-icon-cache does accepts only one icon path

romildo committed:
enlightenment.efl: add missing parent icon themes

romildo commented:
Using symlinks has one complexity: indirect inheritance. It would be needed to symlink also the parent themes from the parents, and so on, up to the inheritance root. This is not implemented yet in this PR.

Also I am not liking the number of collisions.

romildo commented:
cc @worldofpeace

romildo committed:
hicolor-icon-theme: add hook to symlink inherited icon themes

romildo committed:
moka-icon-theme: add hook to symlink inherited icon themes

romildo committed:
arc-icon-theme: add missing parent icon themes

romildo committed:
deepin.deepin-icon-theme: add missing parent icon themes

romildo committed:
elementary-xfce-icon-theme: add missing parent icon themes

romildo committed:
gnome3.adwaita-icon-theme: do not drop icon theme cache

romildo committed:
mate.mate-icon-theme-faenza: add missing parent icon themes

romildo committed:
gnome3.gnome-themes-extra: add missing parent icon themes

romildo committed:
zafiro-icons: add missing parent icon themes

romildo committed:
tango-icon-theme: add missing parent icon themes

romildo committed:
qogir-icon-theme: add missing parent icon themes

romildo committed:
papirus-icon-theme: add missing parent icon themes

romildo committed:
paper-icon-theme: add missing parent icon themes

romildo committed:
numix-icon-theme: add missing parent icon themes

romildo committed:
maia-icon-theme: move postFixup actions to postInstall

romildo committed:
iconpack-obsidian: add missing parent icon themes

romildo committed:
iconpack-jade: add missing parent icon themes

romildo committed:
flat-remix-icon-theme: add missing parent icon themes

romildo committed:
faba-mono-icons: add missing parent icon themes

romildo committed:
faba-icon-theme: add missing parent icon themes

romildo commented:
Now there is a setup hook in `hicolor-icon-themes` that, for each installed icon theme, finds its parent themes by looking at `theme.index`. Then each parent theme is searched among the dependencies. If found the theme directory is linked at `$out/share/icons`.

The dependencies providing parent icon themes should be propagated.

It works for indirect inheritance.

worldofpeace commented:
I like the approach @jtojnar has suggested here, I also really don't favor using `propagatedUserEnv`. But I'm honestly not sure, what was the margin of encapsulation and purity we even wanted out of icon themes in nixpkgs?

> The dependencies providing parent icon themes should be propagated.
> It works for indirect inheritance.
> I have also tried to fix all packages providing icon themes in nixpkgs in order to indicate the dependencies needed to find the parent themes.

I think this is something we should document in the gnome section in the nixpkgs manual.

@jtojnar do you think backporting https://github.com/NixOS/nixpkgs/pull/85229/commits/682734d8526d9cbb273c0c4d0e12aaa6ebb7897f would be a good idea for 20.03?

> Also I am not liking the number of collisions.

These collision warnings have always been annoying, :smile: even if this introduced some.
I don't think there is a way to avoid that, unless we just make it be quiet in general.

romildo commented:
Fixed merge conflicts.

worldofpeace commented:
This is approved @romildo but we need to target staging since this appears to be a mass rebuild.

romildo commented:
> This is approved @romildo but we need to target staging since this appears to be a mass rebuild.

@worldofpeace Targeting staging introduces some file conflicts. I suppose I have to fix them. Any direction on how to do it?

romildo committed:
Merge branch 'staging' into fix.icon-theme-inheritance

nixos-discourse commented:
This pull request has been mentioned on **NixOS Discourse**. There might be relevant details there:

https://discourse.nixos.org/t/proliferation-of-symbolic-links-in-run-current-system-sw-share-icons/7587/1
